feat: add axum http server to interact with interfaces

S-Mann opened the pull request.

S-Mann committed:
✨ Added axum server and server startup to galemind

Signed-off-by: S-Mann <29407727+S-Mann@users.noreply.github.com>

jocct committed:
new file:   mod.rs
	new file:   server.rs

jocct committed:
Separate logic of the server

S-Mann committed:
🧐 Added route handling through a hashmap

Signed-off-by: S-Mann <29407727+S-Mann@users.noreply.github.com>

S-Mann committed:
➕ Added tower_http and hyper

Signed-off-by: S-Mann <29407727+S-Mann@users.noreply.github.com>

jocct committed:
Fixed server

jocct committed:
modified:   server.rs

giorgiozoppi committed:
Fix the handler for the check

giorgiozoppi committed:
Adding compiling

giorgiozoppi committed:
Fixed

giorgiozoppi committed:
Adding v2

giorgiozoppi committed:
Fixnug routers

giorgiozoppi committed:
Fixes on healthcheck

giorgiozoppi committed:
modified:   builder.rs
	modified:   healthcheck.rs
	modified:   mod.rs
	new file:   model.rs

giorgiozoppi committed:
Adding data model to the rest server for inference

giorgiozoppi committed:
Fix the stuff

giorgiozoppi committed:
Adding integration tests

giorgiozoppi committed:
modified:   ../Cargo.lock
	modified:   ../Cargo.toml
	deleted:    api_tests.rs

giorgiozoppi committed:
Adding make run

giorgiozoppi committed:
Add the model

giorgiozoppi commented:
@S-Mann  can you add description This pull request introduces an Axum-based HTTP server to the Galemind project, enabling external interfaces to interact with the system via HTTP endpoints. The integration facilitates communication between the core application and various frontends or tools, such as web dashboards or CLI clients.

Key Features
Axum HTTP Server Integration: Establishes a robust and efficient HTTP server using the Axum framework, known for its performance and type safety in Rust applications.

API Endpoint Implementation: Defines a set of HTTP endpoints that allow external interfaces to perform operations such as data retrieval, updates, and command execution within the Galemind system.

Modular Architecture: The server is designed with modularity in mind, ensuring that additional endpoints and functionalities can be added with minimal effort.

Security Considerations: Implements basic security measures, including input validation and error handling, to ensure reliable and secure interactions.

Impact
By incorporating an HTTP server, Galemind becomes more accessible and versatile, allowing for seamless integration with various user interfaces and tools. This enhancement lays the groundwork for future developments, such as real-time data visualization, remote control capabilities, and integration with third-party services.

Next Steps
Authentication and Authorization: Implement robust authentication mechanisms to secure the HTTP endpoints against unauthorized access.

Comprehensive Testing: Develop unit and integration tests to ensure the reliability and stability of the HTTP server and its endpoints.

Documentation: Provide detailed documentation for the API endpoints, including usage examples and integration guidelines for developers.

This foundational work significantly enhances Galemind's capabilities, promoting greater flexibility and extensibility for future developments.

S-Mann committed:
♻️ Made API version a param

Signed-off-by: S-Mann <29407727+S-Mann@users.noreply.github.com>

giorgiozoppi commented on src/rest_server/builder.rs:
```diff
@@ -0,0 +1,32 @@
+        let addr = addr.parse().expect("Invalid address");
+
+        let app = Router::new()
+            .nest("/{version}/health", new_health_check_router())
```
@S-Mann  perfect!  It is good for manitainability.

giorgiozoppi committed:
modified:   ../../Cargo.lock
	modified:   ../../Cargo.toml
	deleted:    api/inference/mod.rs
	deleted:    api/mod.rs
	deleted:    api/model/mod.rs
	deleted:    api/tensor.rs
	deleted:    config/mod.rs
	deleted:    inference/fake.rs
	deleted:    inference/mod.rs
	deleted:    mod.rs
	deleted:    ../main.rs
	deleted:    ../rest_server/builder.rs
	deleted:    ../rest_server/datamodel.rs
	deleted:    ../rest_server/healthcheck.rs
	deleted:    ../rest_server/mod.rs
	deleted:    ../rest_server/model.rs
 Untracked files:
	../../docs/
	../

giorgiozoppi committed:
new file:   galemind/Cargo.toml
	new file:   galemind/main.rs
	new file:   galemind/src/main.rs
	new file:   rest_server/Cargo.toml
	new file:   rest_server/src/builder.rs
	new file:   rest_server/src/datamodel.rs
	new file:   rest_server/src/healthcheck.rs
	new file:   rest_server/src/lib.rs
	new file:   rest_server/src/mod.rs
	new file:   rest_server/src/model.rs
 Untracked files:

giorgiozoppi committed:
modified:   src/galemind/src/main.rs
	modified:   src/rest_server/src/lib.rs
	modified:   src/rest_server/src/model.rs

giorgiozoppi committed:
Merge branch 'main' of https://github.com/zenforcode/galemind into feature/add-axum-http-server-to-interact-with-interfaces

giorgiozoppi committed:
modified:   Cargo.toml

giorgiozoppi committed:
modified:   Makefile

bcncpp commented:
Added modular arch to be merged before grpc one @S-Mann @S-Mann

S-Mann committed:
⚰️ Removed old builder code

Signed-off-by: S-Mann <29407727+S-Mann@users.noreply.github.com>

S-Mann committed:
🚚 Rename data module for models and moved model endpoints under the models route

Signed-off-by: S-Mann <29407727+S-Mann@users.noreply.github.com>

S-Mann committed:
✨ Added server metadata response

Signed-off-by: S-Mann <29407727+S-Mann@users.noreply.github.com>

bcncpp commented:
gonna resolve conflicts.

S-Mann commented:
I am resolving the conflicts and addressing @markojeremic comments

giorgiozoppi committed:
Merge branch 'main' of https://github.com/zenforcode/galemind into feature/add-axum-http-server-to-interact-with-interfaces

Fixing conflicts

giorgiozoppi commented:
just resolved conflicts. please continue @S-Mann

bcncpp commented:
@S-Mann @markojeremic do you think a template like this for the main works?
``The pattern might be #[tokio::main]
async fn main() {
    let handle1 = tokio::spawn(async {
       //start rest
    });
    let handle2 = tokio::spawn(async {
    ///start grpc.
});
    // Wait for both to complete (will usually run forever)
    let _ = tokio::join!(handle1, handle2);
}
``
In any case systemd will do the start and the stop or do we need to set the pid somewhere?

S-Mann committed:
✨ Added inference req and resp and metadata related APIs

Signed-off-by: S-Mann <29407727+S-Mann@users.noreply.github.com>

S-Mann commented:
> @S-Mann @markojeremic do you think a template like this for the main works? ``The pattern might be #[tokio::main] async fn main() {
> 
> ```
> let handle1 = tokio::spawn(async {
>    //start rest
> });
> let handle2 = tokio::spawn(async {
> ///start grpc.
> ```
> 
> }); // Wait for both to complete (will usually run forever) let _ = tokio::join!(handle1, handle2); } `` In any case systemd will do the start and the stop or do we need to set the pid somewhere?

we plan on running grpc and rust in parallel? I am happy with this, I don't want to make things to generic, will limit functionality or make things types too complex

jocct committed:
Add common interface

jocct committed:
modified:   ../../.devcontainer/devcontainer.json
	modified:   src/lib.rs
	modified:   ../rest_server/src/lib.rs

jocct committed:
- Added common trait InferenceServerBuilder and common config. integrated all in main with one server

jocct committed:
Fixing mames

jocct committed:
Fix metadata model.

jocct committed:
fix formatting

jocct committed:
Fixing error to support async move.

bcncpp commented:
@S-Mann  ok, the test we do in a separate pr. Thank you for the great work and time!
